rental/src/main/java/io/axoniq/demo/bikerental/rental/ui/BikeRentalDataGenerator.java

```diff
@@ -0,0 +1,141 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+@Component
+public class BikeRentalDataGenerator {
+    private static final List<String> RENTERS = Arrays.asList("Allard", "Steven", "Josh", "David", "Marc", "Sara", "Milan", "Jeroen", "Marina", "Jeannot");
+    private static final List<String> LOCATIONS = Arrays.asList("Amsterdam", "Paris", "Vilnius", "Barcelona", "London", "New York", "Toronto", "Berlin", "Milan", "Rome", "Belgrade");
+
+
+    private final CommandGateway commandGateway;
+    private final QueryGateway queryGateway;
+
+    Logger logger = LoggerFactory.getLogger(BikeRentalDataGenerator.class);
+
+    public BikeRentalDataGenerator(CommandGateway commandGateway, QueryGateway queryGateway) {
+        this.commandGateway = commandGateway;
+        this.queryGateway = queryGateway;
+    }
+
+
+    public String randomRenter() {
+        return RENTERS.get(ThreadLocalRandom.current().nextInt(RENTERS.size()));
+    }
+
+    public String randomLocation() {
+        return LOCATIONS.get(ThreadLocalRandom.current().nextInt(LOCATIONS.size()));
+    }
+
+    public CompletableFuture<Void> generateBikes(int bikeCount, String bikeType) {
+        CompletableFuture<Void> all = CompletableFuture.completedFuture(null);
+        for (int i = 0; i < bikeCount; i++) {
+            all = CompletableFuture.allOf(all,
+                    commandGateway.send(new RegisterBikeCommand(UUID.randomUUID().toString(), bikeType, randomLocation())));
+        }
+        return all;
+    }
+
+
+    public Flux<String> generateRentals(String bikeType,
+                                int loops,
+                                int concurrency,
+                                int abandonPaymentFactor,
+                                int delay) {
+
+        return this.internalGenerateRentals(bikeType, loops, concurrency, abandonPaymentFactor, delay);
+
+    }
+
+    private Flux<String> internalGenerateRentals(String bikeType,
+                                 int loops,
+                                 int concurrency,
+                                 int abandonPaymentFactor,
+                                 int delay) {
+
+        return Flux.range(0, loops)
+                .flatMap(j -> executeRentalCycle(bikeType, randomRenter(), abandonPaymentFactor, delay)
+                                .map(r -> "OK - Rented, Payed and Returned\n")
+                                .onErrorResume(e -> Mono.just("Not ok: " + e.getMessage() + "\n")),
+                        concurrency);
+
+    }
+
+    private Mono<String> executeRentalCycle(String bikeType, String renter, int abandonPaymentFactor, int delay) {
+        CompletableFuture<String> result = selectRandomAvailableBike(bikeType)
```

Comment:
I tend to find this way of chaining Completable Futures at different levels a little bit difficult to read. I'd rather have the lambdas defined as separate fields or encapsulated in a method so that it could be easier to read and delimit the different levels.

Also it will make it easier, for example,  to add some logs in the form of "{User} is renting bike {bikeId}" or "{user} is returning bike {id}" which will be more meaningful lines in the logs than the current list of "OK -Rented, Payed and Returned"

Anyway, this is just a comment, we can improve this later.

Reply:
I agree with your thoughts here. Will work to improve this over the next two weeks.